soccer/gameplay/tactics/defensive_forward.py

```diff
@@ -0,0 +1,215 @@
+
+        self.add_transition(
+            DefensiveForward.State.collecting, DefensiveForward.State.blocking,
+            lambda: not self.within_range(), 'Back to blocking')
+
+        self.add_transition(
+            DefensiveForward.State.collecting,
+            behavior.Behavior.State.completed,
+            lambda: self.collector is not None and \
+                    self.collector.robot is not None and \
+                    evaluation.ball.robot_has_ball(self.collector.robot),
+            'Ball collected')
+
+        # Create list of defenders and start the marking
+    def on_enter_blocking(self):
+        self.zone_def_pos, self.mark_bots[0], self.mark_bots[
+            1] = evaluation.defensive_positioning.find_defense_positions()
+
+        self.set_mark_targets()
+
+        self.defenders[2] = skills.mark.Mark()
+        self.add_subbehavior(self.defenders[2], self.names[2], required=True)
+        self.defenders[2].mark_point = self.zone_def_pos
+
+        # Reset time since call
+        self.time_since_call = 0
+
+    # Continue updating the mark positions
+    def execute_blocking(self):
+        self.time_since_call = (self.time_since_call + 1) % self.cache_amnt
+
+        # You don't want to test if its 0 because it would execute instantly
+        if (self.time_since_call == self.cache_amnt - 1):
+            self.zone_def_pos, self.mark_bots[0], self.mark_bots[1] = \
+                evaluation.defensive_positioning.find_defense_positions(
+                    [self.defenders[0].robot, self.defenders[1].robot])
+
+            for i in range(0, 2):
+                if (self.defenders[i].mark_robot is not self.mark_bots[i]):
+                    self.defenders[i].mark_robot = self.mark_bots[i]
+
+            self.defenders[2].mark_point = self.zone_def_pos
+
+    def on_exit_blocking(self):
+        self.remove_all_subbehaviors()
+
+    # Create collector method and then
+    def on_enter_collecting(self):
+        self.zone_def_pos, self.mark_bots[0], self.mark_bots[1] = \
+            evaluation.defensive_positioning.find_defense_positions()
+
+        # Take closest robot to collect
+        # Leave other two to move into primary blocking positions
+        self.collector = skills.capture.Capture()
+        self.add_subbehavior(self.collector, 'collector', required=True)
+
+        self.set_mark_targets()
+
+    def on_exit_collecting(self):
+        self.remove_all_subbehaviors()
+        self.collector = None
+
+    def set_mark_targets(self):
+        for i in range(0, 2):
+            self.defenders[i] = skills.mark.Mark()
+            self.add_subbehavior(self.defenders[i],
+                                 self.names[i],
+                                 required=False,
+                                 priority=20 - 5 * i)
+            self.defenders[i].mark_robot = self.mark_bots[i]
+
+    # Uses their predicted kick direction to block
+    def get_block_pos(self, bot):
```

Comment:
Oh yeah, I completely forgot about this. It was going to be a ghetto version of `mark` before I found out `mark` exist.